### DISCUSSION TITLE
1.Why can't CUDA-accelerated PointCollisionModel be created successfully？2.why the addition of CarvingManager components to CUDA-accelerated TriangleCollisionModel is ineffective?  Are there any good solutions?

### DISCUSSION THREAD

**Discussion by yangwxb** [sofa-framework/sofa#4662] [SofaCUDA]
![1](https://github.com/sofa-framework/sofa/assets/167422897/c3fe2dec-90fd-4b0b-9064-c5ec9e9810cb)
 Object type PointCollisionModel<> was not created
The object is in the factory but cannot be created.
Requested template : CudaVec3f
Used template      : None
Also tried to create the object with the template 'Vec3d' but failed for the following reason(s):
    1. No mechanical state with the datatype 'Vec3d' found in the context node.

**Comment by bakpaul** [comment]
Hello ! 

You might replace the `PointCollisionModel` by `CudaPointCollisionModel`. But in the end this will not speedup the collision detection process: it will still be performed on CPU because no GPU implementation exist. It might be best to map a Vec3 model (CPU DOFS) to your cuda model and add a `PointCollisionModel` to this mapped model. 

Good luck with that !

**Reply by yangwxb** [reply]
Thank you very much for your answer, but I don't quite understand the meaning of the last one sentences:" It might be best to map a Vec3 model (CPU DOFS) to your cuda model and add a PointCollisionModel to this mapped model.", could you give me an example to explain it?
Thanks again!

**Reply by bakpaul** [reply]
The idea is simple to get. Here you are trying to perform collision on the full model. If you have a volumetric object, doing so is sub-optimal because normally, collision can only arise at the surface of the object. Considering the whole object instead of the subset of surface point will increase the collision detection time while having no influence on the actual simulation precision/stability. 

To avoid a lot of unnecessary computation time (again, here this holds for volumetric object, otherwise you will gain nothing by doing so because all the object points are surface points) I suggest to instead use a mapped collision model which will only contain the surface points. Those nodes doesn't need to be on the GPU because it will change nothing in term of collision detection given the fact that it is done on the CPU, and it actually might avoid unnecessary GPU to CPU data transfer which take a lot of time.  

A simple example could be the following : 
```xml
<Node name="GPUObject">
    ...
    <MechanicalObject template="CudaVec3"  />  <!-- Contains all of your model points  -->
    ...
    <Node name="CPUSurface">
        ...
        <MechanicalObject  template="Vec3" /> <!-- Contains only the points of the surface  -->
        <PointCollisionModel />
        ...
        <BarycentricMapping />
    </Node>
</Node>
```

**Reply by yangwxb** [reply]
I am very sorry for replying you so late. I did not explain many things clearly before because I was very unfamiliar with sofa. Now I have an example here. After I add template="CudaVec3", The TriangularFEMForceField<> was not created. Is there a solution to this error?
Thank you again and my best regards!



<VisualStyle displayFlags="showVisual showBehaviorModels" />
    <DefaultAnimationLoop/>
    <CollisionPipeline verbose="0" />
    <BruteForceBroadPhase/>
    <BVHNarrowPhase/>
    <DefaultContactManager response="PenalityContactForceField" />
    <MinProximityIntersection name="Proximity" alarmDistance="0.8" contactDistance="0.5" />
    <Node name="SquareGravity">
        <EulerImplicitSolver name="cg_odesolver" printLog="false"  rayleighStiffness="0.1" rayleighMass="0.1" />
        <CGLinearSolver iterations="25" name="linear solver" tolerance="1.0e-9" threshold="1.0e-9" />
        <MeshOBJLoader name="meshLoader" filename="mesh/square_2594_triangles.obj" scale="10" createSubelements="true" />
        <TriangleSetTopologyContainer  name="Container" src="@meshLoader"/>
        <TriangleSetTopologyModifier   name="Modifier" />
        <TriangleSetGeometryAlgorithms name="GeomAlgo" template="CudaVec3f" />
        <MechanicalObject name="Mo" template="CudaVec3f"/>
        <DiagonalMass massDensity="0.08" />
        <FixedConstraint indices="617 618 57 1301 1302 49" />
        <TriangularFEMForceField name="FEM" youngModulus="6000" poissonRatio="0.3" method="large" template="CudaVec3f" />
        
        <TriangleCollisionModel />

**Reply by yangwxb** [reply]
![Uploading 11.jpg…]()

**Reply by bakpaul** [reply]
The cuda implementation of `TriangulaFEMForcefield` is called `TriangularFEMForceFieldOptim` could you try this to see if it works. 

For the collision model I aready gave you an answer. But if you still want to have a collision model on your GPU dofs, you'll need to use the `CudaXXXCollisionModel` by replacing `XXX` by the type of ollision you want in the available ones : `{Point, Line, Sphere, Triangle}`

**Reply by yangwxb** [reply]
![1](https://github.com/sofa-framework/sofa/assets/167422897/49c38b21-fcb0-4738-b0cf-9f22fc2433b9)
![2](https://github.com/sofa-framework/sofa/assets/167422897/4e0eb6ad-4255-4ccd-9a61-d0ccbb630d80)
Can you see my picture messages? I tried to follow your advice, but it didn't work.

**Reply by yangwxb** [reply]
Sorry to bother you, but I am now faced with an incomprehensible phenomenon: when I added a model to the original scn file, the sofa effect of the original model changed a lot, and I did not change any of its parameters.What is the reason for this?
